### ISSUE TITLE
:Gcommit incompatible with 'set autochdir'

### ISSUE THREAD

**Issue by ChrisPenner** [tpope/vim-fugitive#1195]
When I open `:Gstatus` and run `:Gcommit` I get:

```
fugitive: fatal: This operation must be run in a work tree
```

My `:pwd` in the `:Gstatus` window shows that I'm inside the `project-root/.git` directory, so it makes sense why it's failing, I suspect this is due to the fact that I have `set autochdir` in my vim config.

Unfortunately this isn't so easy to fix, I tried a simple `lcd ../` in a `BufEnter` autocmd, but it seems that the `autochdir` happens AFTER that, I'm not aware of any autocmd triggers that happen after that.

Have any ideas? Perhaps `:Gcommit` could be run from the git root instead of `pwd`? If there's anything I can change in my config I'm happy to.

Thanks for the great plugins!

**Comment by tpope** [comment]
It *does* force the Git root, and it's working fine for me even with `'autochdir'`. Methinks you have something else nefarious in your config.

**Comment by ChrisPenner** [comment]
Interesting! I'll take another look

**Comment by ptrin** [comment]
I also have this issue. It seems that in the `:Gstatus` buffer my working directory is `.git`. Running `:cd ..` made `:Gcommit` work again.

**Comment by tpope** [comment]
Is this a new issue, or has the `:Gstatus` overhaul merely drawn your attention to it?

**Comment by ptrin** [comment]
I think it's probably a new issue. If it helps to debug, my vimrc is here: https://github.com/ptrin/vimconfig/blob/master/vimrc

**Comment by tpope** [comment]
Try on version 2.5, please, "probably" isn't very helpful.

**Comment by ChrisPenner** [comment]
The issue exists on version 2.5 (for me at least)

**Comment by tpope** [comment]
Sounds like a plausible hypothesis.

**Comment by ptrin** [comment]
I don't think it's related to neovim because I don't use it but still have the issue. For now I've disabled `autochdir`.

**Comment by ltgouvea** [comment]
I started experiencing this issue when I updated/recompiled my vim 8 so I could install [coc.nvim](https://github.com/neoclide/coc.nvim). Disabling `autochdir` made everything work normally again.

**Comment by tpope** [comment]
Has anyone tried reproducing the issue with other plugins disabled and a minimal vimrc?

If you edit a file, `:cd` to a different directory, and call `:!pwd`, does it report the new directory or the parent directory of the file?

**Comment by ltgouvea** [comment]
I can try that scenario, @tpope, how specific the vimrc should be?

**Comment by tpope** [comment]
`:set autochdir` plus whatever you need to get it to load Fugitive.

That second question can also be answered outside of the minimal vimrc context.

**Comment by ltgouvea** [comment]
I successfully reproduced the issue using only `:autochdir` + Fugitive installed. When i used `:cd` the directory reported was the new directory.

**Comment by tpope** [comment]
Here's the relevant code. It looks complicated but most of it is just building the argument string. The `s:Cd()` call changes to the working directory to the work tree. I guess my recommendation would be to throw a bunch of `:pwd`s in there and see if the working directory is correct.

https://github.com/tpope/vim-fugitive/blob/41a734ef49c78cfdb97426cda811ee99c39e011a/autoload/fugitive.vim#L2665-L2690

**Comment by tpope** [comment]
We now use `git -C <worktree>`, so this issue *should* be moot.

**Comment by sim590** [comment]
I still experience this issue while using version 5eeb7625df75a0da931d81a84e8730aca8b8c8a1 (2019-10-01). The value of  `pwd` is `.git` in `:Gstatus` window and my work tree when in my work buffers. I do use `autochdir` option.

**Comment by tpope** [comment]
Add `:echo command` before this line and see what it's trying to run: https://github.com/tpope/vim-fugitive/blob/80aa8a9aea72f2054fb2ad8582461f60b0db9757/autoload/fugitive.vim#L3298

**Comment by sim590** [comment]
@tpope I had to put `:echoerr` instead to see the message. Otherwise, the message would not show up. I get this:

```
env GIT_EDITOR=false git commit
```

**Comment by tpope** [comment]
That `command` is generated by `s:UserCommand()` a few lines above, which in turn is a thin wrapper around `s:UserCommandList()`, which does the following:

https://github.com/tpope/vim-fugitive/blob/698065596c9ccee74825ac9667357ce8222ed7b0/autoload/fugitive.vim#L245-L247

In English, this is "if there is a non-empty work tree (not a bare repository), and that work tree is not the current directory, add a `-C` argument". So this should trigger if the working directory is `.git` as you suggest.

You can try adding `echoerr tree . ',' . getcwd()` there and see if that matches your expectations.

**Comment by sim590** [comment]
@tpope the output of `tree` is the root of the working tree, but `pwd` is `$ROOT/.git`. Therefore, there's something wrong because I still get `fatal: this operation must be run in a work tree` error message.

**Comment by sim590** [comment]
I have just found out something useful: opening the file directly using `vim myfile`, then I don't have the bug. If I open only with command `vim` and select the file using `:Startify` (from startify plugin buffer), then I get the bug! I'm sorry that I didn't try without disabling everything. I know that I made you work may be for nothing. But is that useful information? Could vim-fugitive try to protect against whatever is causing this? Otherwise, may be there's something I could report to Startify upstream?

**Comment by tpope** [comment]
> @tpope the output of `tree` is the root of the working tree, but `pwd` is `$ROOT/.git`. Therefore, there's something wrong because I still get `fatal: this operation must be run in a work tree` error message.

So I guess try adding an `echoerr` inside that `elseif` to see if it triggers. And then check the value of `git`, etc. *Something* isn't adding up.



> I have just found out something useful: opening the file directly using `vim myfile`, then I don't have the bug. If I open only with command `vim` and select the file using `:Startify` (from startify plugin buffer), then I get the bug! I'm sorry that I didn't try without disabling everything. I know that I made you work may be for nothing. But is that useful information? Could vim-fugitive try to protect against whatever is causing this? Otherwise, may be there's something I could report to Startify upstream?

Fugitive could possibly guard against something, but it's on you to find what that something is. Little harm in opening a Startify issue, but `'autochdir'` is a cross you choose to bear, it's still probably going to have to be you that finds the problem.

**Comment by sim590** [comment]
Actually, Startify may have little to do with this. I still get the bug for some reason after some manipulation that I don't recall (but without using Startify). I fixed by issuing `:cd .` ... :/

Anyway, when opening with Startify while `echoerr` is placed in the `elseif` you mentioned, there's no output, so it's not called. I tried with Startify because it seems that I can trigger the bug like this. As mentioned in my last paragraph, I still can trigger the bug randomly without, but I'm not sure how.
